### PR TITLE
Added Merkle tree

### PR DESCRIPTION
**Pull request by amitkumarj441**
Thanks for the pull request!
Pull requests should have these details:

## What is this pull request?
Add Merkle Signature Option #3 
   
## Why do you implement it? Why do we need this pull request?
As MSS can be quantum resistant, so adding these to Iroha would be beneficial.

### CODE REVIEW

**Comment by neewy** [comment]
@Warchant there is a problem with sonar scanner, which worth checking after PR review

File: data.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    // Encodes byte array of object into base64 string
+    std::string toString();
+    // Returns the size of the data object
+    size_t size();
+    byte bytes[kDigestSize];
+
+    /*------ Static functions ------*/
+    // Hashes a message to size bytes: default 32 bytes
+    static Data hashMessage(std::string message, int messageLen, int size=kDigestSize);
```
Comment (l4l) [inline data.h]: `message.size() == messageLen`? second arg seem redundant

File: libs/crypto/merkle.cpp [hunk 2]
```diff
@@ -0,0 +1,273 @@
+#include <sstream>
+#include <iostream>
+
+#include "merkle.h"
+using IROHA_CRYPTO_HPP::Integer;
+
+#define BLOCKSIZE 16
+#define DIGESTSIZE 32
+
+using namespace std;
```
Comment (l4l) [inline libs/crypto/merkle.cpp]: oh, quite dangerous, don't use this please

File: libs/crypto/merkle.cpp [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+  merkSig.auth = getAuthPath(msgNum);
+  merkSig.msgNum = msgNum;
+  return merkSig;
+}
+
+/*
+ * In the subclass, this is a critical region.
+ */
+long Merkle::getNextMsg() {
+  msg++;
```
Comment (l4l) [inline libs/crypto/merkle.cpp]: quite bad to have side-effects on functions with get-names

File: merkle.h [hunk 4]
```diff
@@ -0,0 +1,113 @@
+#include "winternitz.h"
+#include <stack>
+
+class Merkle {
+  public:
+    struct Signature {
+      std::vector<Data> wint; 
+      std::vector<Data> auth;
```
Comment (l4l) [inline merkle.h]: what does these vectors mean/represent?

File: merkle.h [hunk 5]
```diff
@@ -0,0 +1,113 @@
+#include "winternitz.h"
```
Comment (l4l) [inline merkle.h]: "winternitz"? :o

File: data.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    // Empty constructor
+    Data();
+    // Converts byte array into Data object
+    Data(byte* data, size_t sizeIn = kBlockSize);
+    // Converts Integer into Data object
+    Data(IROHA_CRYPTO_HH::Integer data, unsigned int size = kBlockSize);
+    ~Data();
+
+    // Encodes byte array of object into base64 string
+    std::string toString();
```
Comment (l4l) [inline data.h]: imo it's better to make it `static std::string to_string(const Data&)`, just as in stl

**Comment by amitkumarj441** [comment]
Yes @neewy, seems Sonar Scanner is unhappy with my PR :disappointed:

File: data.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
+#include <iostream>
+#include <stdio.h>
+#include <stdlib.h>
```
[the remark this answers is not in the document]
Comment (amitkumarj441) [reply data.h]: Thanks @l4l for the reminder.
I just updated it.

File: data.h (hunk 6 above)
Comment (amitkumarj441) [reply data.h]: Alright :+1:

File: data.h (hunk 1 above)
Comment (amitkumarj441) [reply data.h]: Thanks @l4l for pointing this.

File: libs/crypto/merkle.cpp (hunk 2 above)
Comment (amitkumarj441) [reply libs/crypto/merkle.cpp]: Okay, got it. Any alternative for this?

Comment (Warchant) [reply libs/crypto/merkle.cpp]: if you need something specific, for example `std::vector`

use:
```
using std::vector
```

**Comment by Warchant** [comment]
Don't forget to rebase on `develop`, it seems I fixed the issue with `sonar-pr`. 

Don't forget to sign DCO with **every** commit:
```
git commit -s -m  "commit message"
```

File: libs/crypto/merkle.cpp (hunk 2 above)
Comment (amitkumarj441) [reply libs/crypto/merkle.cpp]: Yes @Warchant , I think `#include <vector>` will do this.

**Comment by amitkumarj441** [comment]
@Warchant Sorry, I mistakenly ran `git push -f` and when signing with `git commit -s -m  "commit message"` it's throwing 
```
ith@ith-ThinkPad-W520:~/Practice/iroha$ git commit -s -m "Signed-off-by: Amit Kumar Jaiswal <amitkumarj441@gmail.com>"
On branch patch-1
nothing to commit, working directory clean
```
Can you please deleter those unrelated commits except my changes in PR?

**Comment by Warchant** [comment]
Sure.

**Comment by Warchant** [comment]
:face_with_head_bandage: it seems I did something wrong...
 
Please, go to your local repo and:
```bash
# fetch latest develop. assuming that hyperledger/iroha is upstream remote
$ git fetch upstream
$ git checkout patch-1
$ git rebase -i upstream/develop
# then remove all lines with commits that you want to delete
$ git push -f origin patch-1
```

**Comment by amitkumarj441** [comment]
Thanks @Warchant for removing irrelevant commits to this PR. But commit [c9d67e9](https://github.com/hyperledger/iroha/pull/631/commits/c9d67e93931247152a2fa1dbe415159fbf8df9f9) needs sign-off.

**Comment by l4l** [comment]
Well, you actually don't required to make real signing. Comment with __Signed-off-by__ is enough
You can try [this](https://gist.github.com/l4l/694ca936ef538a881cdc3604cc12b2e0) script for make this automatically (though it may have bug(s))

**Comment by Warchant** [comment]
P.S. you don't have to **cryptographically** sign commits. You need to sign DCO. Every commit should contain 

```
Signed-off-by: Kitsu <mail@kitsu.me>
```

```
$ git commit
(your favourite editor will be open)
<1st line: commit description>
<blank line>
Signed-off-by: Kitsu <mail@kitsu.me>
<save file>
```
Done! You just signed it. 
Easy way is to predefine a template or to use -s:
```
$ git commit -s
```

**Comment by amitkumarj441** [comment]
Thanks @l4l and @Warchant for letting me know about PRs rules .
Sorry for polluting my PRs and could you remove this irrelevant [commit](https://github.com/hyperledger/iroha/pull/631/commits/0ff9d0b5cfddb88c0e0704e3e11f15bac914c81a).
Out of the six main commits, I signed 2 of them and 4 of them still unsigned.
Please let me know should I sign my other 4 commits ( as @Warchant already signed it and that's why I got confused ).

**Comment by neewy** [comment]
@amitkumarj441 hi! What is the status?

**Comment by amitkumarj441** [comment]
@neewy Have a look at my previous comment? and I guess we're done with review but need to remove this [commit](https://github.com/hyperledger/iroha/pull/631/commits/0ff9d0b5cfddb88c0e0704e3e11f15bac914c81a)

**Comment by neewy** [comment]
Please, try to answer all the questions @l4l asked you. Also, to merge PR we have to get at least another confirmation from Iroha maintainers team.

File: merkle.h (hunk 5 above)
Comment (amitkumarj441) [reply merkle.h]: @l4l The signatures still need a lot of storage and Winternitz involves iterating hashes. Sorry I forgot to push `winternitz.h`. I'll push it soon.

File: merkle.h (hunk 4 above)
Comment (amitkumarj441) [reply merkle.h]: @l4l `std::vector<Data> wint` represents winternitz signature and `std::vector<Data> auth` represents authorization path.

File: merkle.h (hunk 5 above)
Comment (l4l) [reply merkle.h]: ah, nope, nvm
Just never heard of Wintermitz OTSS

File: libs/crypto/merkle.cpp (hunk 3 above)
Comment (amitkumarj441) [reply libs/crypto/merkle.cpp]: Thanks @l4l for pointing this out. But any alternative for get-names?

**Comment by amitkumarj441** [comment]
@neewy Sorry, I just replied on @l4l questions.

File: merkle.h (hunk 5 above)
Comment (amitkumarj441) [reply merkle.h]: @l4l Ohh...I see, I'll have the code ready and now push it here in this PR for review.

File: merkle.h (hunk 4 above)
Comment (l4l) [reply merkle.h]: Do you mind to add those comments directly in the code?

File: libs/crypto/merkle.cpp (hunk 3 above)
Comment (l4l) [reply libs/crypto/merkle.cpp]: Maybe just `nextMsg` ?

File: data.h [hunk 8]
```diff
@@ -0,0 +1,45 @@
+#include <iostream>
+#include <cstdio.h>
+#include <cstdlib.h>
+
+#include "cryptopp/sha.h"
+#include "cryptopp/Integer.h"
+
+class Data {
```
Comment (l4l) [inline data.h]: For crypto data we're using either `std::string` or [`blob_t`](https://github.com/hyperledger/iroha/blob/develop/libs/common/types.hpp#L57)
What're the major features of this class that cannot be done via listed ones?

File: merkle.h [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    void combine();
+
+    // Returns the authorization path for the current message
+    std::vector<Data> getAuthPath(long msgNum);
+
+    // Returns the current message number and increments it
+    virtual long getNextMsg();
+};
+
+class messagesException: public std::exception {
```
Comment (l4l) [inline merkle.h]: we're trying to avoid exceptions

**Comment by neewy** [comment]
@amitkumarj441 please, implement the changes requested. Ask your help at any time from maintainers team to explain you what is needed here or in rocker.chat.

Thanks for your interest to the project!

File: merkle.h (hunk 9 above)
Comment (takemiyamakoto) [reply merkle.h]: You should explain how to handle error cases, then.

**Comment by neewy** [comment]
@amitkumarj441 please respond here about the intention to fix changes. According to https://github.com/hyperledger/iroha/wiki/Iroha-working-agreement:

> 4.10.4 Outdated flow
> 
> {needs review, needs correction, …, candidate for closing}
> 
> In case when:
> 
> pull request is opened for single month without any activity, like review or discussions, or
> review corrections are not fixed for 14 days The pull request is a candidate to be closed, if after change of status and request for the opener to clarify the status no activity happens for 24 hours.

**Comment by amitkumarj441** [comment]
@neewy I'll ping reviewers on rocketchat or gitter to short out this issues and move ahead toward goal of this PR.

Apologies for the delay :disappointed:

**Pull request closed without merging**